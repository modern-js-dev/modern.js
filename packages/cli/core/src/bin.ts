#!/usr/bin/env node
require('../compiled/v8-compile-cache');

// eslint-disable-next-line import/first
import { cli } from '.';

if (!process.env.NODE_ENV) {
  const command = process.argv[2];

  if (['build', 'start', 'deploy'].includes(command)) {
    process.env.NODE_ENV = 'production';
  } else if (command === 'test') {
    process.env.NODE_ENV = 'test';
  } else {
    process.env.NODE_ENV = 'development';
  }
}

const { version } = require('../package.json');

cli.run(process.argv.slice(2), { version });
