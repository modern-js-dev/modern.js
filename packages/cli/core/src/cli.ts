// 这个文件跟 bin/modern-js.js 基本一样
// 在开发阶段，因为 package.json 的 exports['./bin']['jsnext:source'] 配置
// 了这个文件，所以需要保留, 后续如果找到更好的方式之后会移除这个文件

import { cli } from '.';

const { version } = require('../package.json');

process.env.MODERN_JS_VERSION = version;
if (!process.env.NODE_ENV) {
  process.env.NODE_ENV =
    // eslint-disable-next-line no-nested-ternary
    ['build', 'start', 'deploy'].includes(process.argv[2])
      ? 'production'
      : process.argv[2] === 'test'
      ? 'test'
      : 'development';
}

cli.run(process.argv.slice(2));
