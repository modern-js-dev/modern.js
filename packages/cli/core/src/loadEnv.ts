import fs from 'fs';
import path from 'path';
import dotenv from 'dotenv';
import dotenvExpand from 'dotenv-expand';

export const loadEnv = (
  appDirectory: string,
  mode: string = process.env.NODE_ENV as string,
) => {
  [`.env.${mode}`, '.env']
    .map(name => path.resolve(appDirectory, name))
    .filter(
      filePath =>
        fs.existsSync(filePath) && !fs.statSync(filePath).isDirectory(),
    )
    .forEach(filePath => {
      const envConfig = dotenv.config({ path: filePath });

      dotenvExpand(envConfig);
    });
};
