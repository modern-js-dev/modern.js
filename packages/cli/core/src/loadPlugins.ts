import {
  isDepExists,
  createDebugger,
  compatRequire,
  INTERNAL_PLUGINS,
} from '@modern-js/utils';

const debug = createDebugger('load-plugins');

export interface PluginConfigItem {
  cli?: string;
  server?: string;
}

export type PluginConfig = Array<PluginConfigItem>;

/**
 * Try to resolve plugin entry file path.
 * @param appDirectory - Application root directory.
 * @param plugin - Plugin name or plugin name with options.
 * @returns Resolved file path.
 */
const resolvePlugin = (appDirectory: string, plugin: PluginConfigItem) => {
  const tryResolve = (name: string) => {
    let filePath = '';
    try {
      filePath = require.resolve(name, { paths: [appDirectory] });
      delete require.cache[filePath];
    } catch (err) {
      if ((err as any).code === 'MODULE_NOT_FOUND') {
        throw new Error(`Can not find plugin ${name}.`);
      }
      throw err;
    }
    return filePath;
  };

  const resolved: PluginConfigItem = {};

  if (typeof plugin === 'string' || plugin.cli) {
    resolved.cli =
      typeof plugin === 'string' ? tryResolve(plugin) : tryResolve(plugin.cli!);
  }

  if (plugin.server) {
    resolved.server = tryResolve(plugin.server);
  }

  return resolved;
};

/**
 * Load internal plugins which in @modern-js scope and user's custom plugins.
 * @param appDirectory - Application root directory.
 * @param pluginsConfig - Plugins declared in the user configuration.
 * @returns Plugin Objects has been required.
 */
export const loadPlugins = (
  appDirectory: string,
  pluginConfig: PluginConfig,
  internalPlugins?: typeof INTERNAL_PLUGINS,
) => {
  const plugins = [
    ...Object.keys(internalPlugins || INTERNAL_PLUGINS)
      .filter(name => isDepExists(appDirectory, name))
      .map(name => (internalPlugins || INTERNAL_PLUGINS)[name]),
    ...pluginConfig,
  ];

  return plugins.map(plugin => {
    const { cli, server } = resolvePlugin(appDirectory, plugin);

    debug(`resolve plugin %s: %s`, plugin, {
      cli,
      server,
    });

    return {
      cli: cli && compatRequire(cli),
      server: server && compatRequire(server),
    };
  });
};
