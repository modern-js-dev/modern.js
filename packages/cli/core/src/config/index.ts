import { loadConfig } from '@modern-js/load-config';
import Ajv, { ErrorObject } from 'ajv';
import ajvKeywords from 'ajv-keywords';
import logger from 'signale';
import {
  createDebugger,
  getPort,
  isDev,
  MetaOptions,
  PLUGIN_SCHEMAS,
  chalk,
} from '@modern-js/utils';
import mergeWith from 'lodash.mergewith';
import betterAjvErrors from 'better-ajv-errors';
import { codeFrameColumns } from '@babel/code-frame';
import { PluginConfig } from '../loadPlugins';
import { repeatKeyWarning } from '../utils/repeatKeyWarning';
import { defaults } from './defaults';
import { mergeConfig, NormalizedConfig } from './mergeConfig';
import { patchSchema, PluginValidateSchema } from './schema';

const debug = createDebugger('resolve-config');

export { defaults as defaultsConfig };
export { mergeConfig };

interface SourceConfig {
  entries?: Record<
    string,
    | string
    | {
        entry: string;
        enableFileSystemRoutes?: boolean;
        disableMount?: boolean;
      }
  >;
  disableDefaultEntries?: boolean;
  entriesDir?: string;
  configDir?: string;
  apiDir?: string;
  envVars?: Array<string>;
  globalVars?: Record<string, string>;
  alias?:
    | Record<string, string>
    | ((aliases: Record<string, string>) => Record<string, unknown>);
  moduleScopes?:
    | Array<string | RegExp>
    | ((scopes: Array<string | RegExp>) => void)
    | ((scopes: Array<string | RegExp>) => Array<string | RegExp>);
  include?: Array<string | RegExp>;
}

interface OutputConfig {
  assetPrefix?: string;
  htmlPath?: string;
  jsPath?: string;
  cssPath?: string;
  mediaPath?: string;
  path?: string;
  title?: string;
  titleByEntries?: Record<string, string>;
  meta?: MetaOptions;
  metaByEntries?: Record<string, MetaOptions>;
  inject?: 'body' | 'head' | boolean;
  injectByEntries?: Record<string, 'body' | 'head' | boolean>;
  mountId?: string;
  favicon?: string;
  faviconByEntries?: Record<string, string | undefined>;
  copy?: Array<Record<string, unknown> & { from: string }>;
  scriptExt?: Record<string, unknown>;
  disableTsChecker?: boolean;
  disableHtmlFolder?: boolean;
  disableCssModuleExtension?: boolean;
  disableCssExtract?: boolean;
  enableCssModuleTSDeclaration?: boolean;
  disableMinimize?: boolean;
  enableInlineStyles?: boolean;
  enableInlineScripts?: boolean;
  disableSourceMap?: boolean;
  disableInlineRuntimeChunk?: boolean;
  disableAssetsCache?: boolean;
  enableLatestDecorators?: boolean;
  polyfill?: 'off' | 'usage' | 'entry' | 'ua';
  dataUriLimit?: number;
  templateParameters?: Record<string, unknown>;
  templateParametersByEntries?: Record<
    string,
    Record<string, unknown> | undefined
  >;
  cssModuleLocalIdentName?: string;
  enableModernMode?: boolean;
  federation?: boolean;
  disableNodePolyfill?: boolean;
  enableTsLoader?: boolean;
}

interface ServerConfig {
  routes?: Record<
    string,
    | string
    | {
        route: string | string[];
        disableSpa?: boolean;
      }
  >;
  publicRoutes?: { [filepath: string]: string };
  ssr?: boolean | Record<string, unknown>;
  ssrByEntries?: Record<string, boolean | Record<string, unknown>>;
  baseUrl?: string | Array<string>;
  port?: number;
  logger?: boolean | Record<string, any>;
  metrics?: Record<string, any>;
  enableMicroFrontendDebug?: boolean;
}

interface DevConfig {
  assetPrefix?: string | boolean;
  https?: boolean;
}

interface MicroFrontend {
  enableHtmlEntry?: boolean;
  externalBasicLibrary?: boolean;
  moduleApp?: string;
}

interface DeployConfig {
  microFrontend?: false | MicroFrontend;
  domain?: string | Array<string>;
  domainByEntries?: Record<string, string | Array<string>>;
}

type ConfigFunction =
  | Record<string, unknown>
  // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
  | ((config: Record<string, unknown>) => Record<string, unknown> | void);
interface ToolsConfig {
  webpack?: ConfigFunction;
  babel?: ConfigFunction;
  autoprefixer?: ConfigFunction;
  postcss?: ConfigFunction;
  styledComponents?: ConfigFunction;
  lodash?: ConfigFunction;
  devServer?: Record<string, unknown>;
  tsLoader?: ConfigFunction;
  terser?: ConfigFunction;
  minifyCss?: ConfigFunction;
  esbuild?: Record<string, unknown>;
}

type RuntimeConfig = Record<string, any>;

interface RuntimeByEntriesConfig {
  [name: string]: RuntimeConfig;
}

interface UserConfig {
  source?: SourceConfig;
  output?: OutputConfig;
  server?: ServerConfig;
  dev?: DevConfig;
  deploy?: DeployConfig;
  tools?: ToolsConfig;
  plugins?: PluginConfig;
  runtime?: RuntimeConfig;
  runtimeByEntries?: RuntimeByEntriesConfig;
}

type ConfigParam =
  | UserConfig
  | Promise<UserConfig>
  | ((env: any) => UserConfig | Promise<UserConfig>);

interface LoadedConfig {
  config: UserConfig;
  filePath: string | false;
  dependencies: string[];
  pkgConfig: UserConfig;
  jsConfig: UserConfig;
}

export const defineConfig = (config: ConfigParam): ConfigParam => config;

export const loadUserConfig = async (
  appDirectory: string,
  filePath?: string,
  packageJsonConfig?: string,
): Promise<LoadedConfig> => {
  const loaded = await loadConfig<ConfigParam>(
    appDirectory,
    filePath,
    packageJsonConfig,
  );

  const config = !loaded
    ? {}
    : await (typeof loaded.config === 'function'
        ? loaded.config(0)
        : loaded.config);

  return {
    config: mergeWith({}, config || {}, loaded?.pkgConfig || {}),
    jsConfig: (config || {}) as any,
    pkgConfig: (loaded?.pkgConfig || {}) as UserConfig,
    filePath: loaded?.path,
    dependencies: loaded?.dependencies || [],
  };
};

const showAdditionalPropertiesError = (error: ErrorObject) => {
  if (
    error.keyword === 'additionalProperties' &&
    error.params.additionalProperty
  ) {
    const target = [
      error.instancePath.slice(1),
      error.params.additionalProperty,
    ]
      .filter(Boolean)
      .join('.');

    const name = Object.keys(PLUGIN_SCHEMAS).find(key =>
      (PLUGIN_SCHEMAS as Record<string, any>)[key].some(
        (schemaItem: any) => schemaItem.target === target,
      ),
    );

    if (name) {
      logger.warn(
        `The configuration of ${chalk.bold(
          target,
        )} is provided by plugin ${chalk.bold(name)}. Please use ${chalk.bold(
          'yarn new',
        )} to enable the corresponding capability.\n`,
      );
    }
  }
};

/* eslint-disable max-statements, max-params */
export const resolveConfig = async (
  loaded: LoadedConfig,
  configs: UserConfig[],
  schemas: PluginValidateSchema[],
  restartWithExistingPort: number,
  argv: string[],
  onSchemaError: (error: ErrorObject) => void = showAdditionalPropertiesError,
): Promise<NormalizedConfig> => {
  const { config: userConfig, jsConfig, pkgConfig } = loaded;

  const ajv = new Ajv({ $data: true, strict: false });

  ajvKeywords(ajv);

  const validateSchema = patchSchema(schemas);

  const validate = ajv.compile(validateSchema);

  repeatKeyWarning(validateSchema, jsConfig, pkgConfig);

  // validate user config.
  const valid = validate(userConfig);

  if (!valid && validate.errors?.length) {
    onSchemaError(validate?.errors[0]);
    const errors = betterAjvErrors(
      validateSchema,
      userConfig,
      validate.errors?.map(e => ({
        ...e,
        dataPath: e.instancePath,
      })),
      {
        format: 'js',
        indent: 2,
      },
    );

    logger.log(
      codeFrameColumns(
        JSON.stringify(userConfig, null, 2),
        {
          start: errors?.[0].start as any,
          end: errors?.[0].end as any,
        },
        {
          highlightCode: true,
          message: errors?.[0].error,
        },
      ),
    );
    throw new Error(`Validate configuration error`);
  }

  // validate config from plugins.
  for (const config of configs) {
    if (!validate(config)) {
      logger.error(validate.errors);
      throw new Error(`Validate configuration error.`);
    }
  }
  const resolved = mergeConfig([defaults as any, ...configs, userConfig]);

  resolved._raw = loaded.config;

  if (isDev() && argv[0] === 'dev') {
    if (restartWithExistingPort > 0) {
      // dev server is restarted, should use existing port number
      resolved.server.port = restartWithExistingPort;
    } else {
      // get port for new dev server
      resolved.server.port = await getPort(resolved.server.port!);
    }
  }

  debug('resolved %o', resolved);

  return resolved;
};
/* eslint-enable max-statements, max-params */

export type {
  SourceConfig,
  OutputConfig,
  ServerConfig,
  DevConfig,
  DeployConfig,
  ToolsConfig,
  RuntimeConfig,
  RuntimeByEntriesConfig,
  UserConfig,
  ConfigParam,
  LoadedConfig,
};
