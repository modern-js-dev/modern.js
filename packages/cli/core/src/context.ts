import path from 'path';
import { createContext } from '@modern-js/plugin';
import address from 'address';
import type { IAppContext } from '@modern-js/types';
import { UserConfig } from './config';
import { NormalizedConfig } from './config/mergeConfig';

export type { IAppContext };

export const AppContext = createContext<IAppContext>({} as IAppContext);

export const ConfigContext = createContext<UserConfig>({} as UserConfig);

export const ResolvedConfigContext = createContext<NormalizedConfig>(
  {} as NormalizedConfig,
);

export const useAppContext = () => AppContext.use().value;

export const useConfigContext = () => ConfigContext.use().value;

export const useResolvedConfigContext = () => ResolvedConfigContext.use().value;

export const initAppContext = (
  appDirectory: string,
  plugins: Array<{
    cli: any;
    server: any;
  }>,
  configFile: string | false,
  options?: {
    srcDir?: string;
    distDir?: string;
    sharedDir?: string;
    internalDir?: string;
  },
): IAppContext => {
  const {
    srcDir = 'src',
    distDir = '',
    sharedDir = 'shared',
    internalDir = '.modern-js',
  } = options || {};

  return {
    appDirectory,
    configFile,
    ip: address.ip(),
    port: 0,
    packageName: require(path.resolve(appDirectory, './package.json')).name,
    srcDirectory: path.resolve(appDirectory, srcDir),
    distDirectory: distDir,
    sharedDirectory: path.resolve(appDirectory, sharedDir),
    nodeModulesDirectory: path.resolve(appDirectory, './node_modules'),
    internalDirectory: path.resolve(
      appDirectory,
      `./node_modules/${internalDir}`,
    ),
    plugins,
    htmlTemplates: {},
    serverRoutes: [],
    entrypoints: [],
  };
};
