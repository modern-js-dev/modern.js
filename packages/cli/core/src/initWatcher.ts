import crypto from 'crypto';
import fs from 'fs';
import path from 'path';
import { isDev, createDebugger, isTest } from '@modern-js/utils';
import chokidar from 'chokidar';
import { LoadedConfig } from './config';
import { HooksRunner } from '.';

const debug = createDebugger('watch-files');

const md5 = (data: string) =>
  crypto.createHash('md5').update(data).digest('hex');

const hashMap = new Map<string, string>();

export const initWatcher = async (
  loaded: LoadedConfig,
  appDirectory: string,
  configDir: string | undefined,
  hooksRunner: HooksRunner,
  argv: string[],
  // eslint-disable-next-line consistent-return
) => {
  // only add fs watcher on dev mode.
  if ((isDev() || isTest()) && argv[0] === 'dev') {
    const extraFiles = await hooksRunner.watchFiles();

    const configPath = path.join(appDirectory, configDir!);

    const watched = [
      `${configPath}/html`,
      ...(extraFiles as any),
      loaded?.filePath,
      ...loaded.dependencies,
    ].filter(Boolean);

    debug(`watched: %o`, watched);

    const watcher = chokidar.watch(watched, {
      cwd: appDirectory,
      ignorePermissionErrors: true,
      ignored: [
        /node_modules/,
        '**/__test__/**',
        '**/*.test.(js|jsx|ts|tsx)',
        '**/*.spec.(js|jsx|ts|tsx)',
        '**/*.stories.(js|jsx|ts|tsx)',
      ],
    });

    watcher.on('change', changed => {
      const lastHash = hashMap.get(changed);
      const currentHash = md5(
        fs.readFileSync(path.join(appDirectory, changed), 'utf8'),
      );
      if (currentHash !== lastHash) {
        debug(`file change: %s`, changed);

        hashMap.set(changed, currentHash);

        hooksRunner.fileChange({ filename: changed, eventType: 'change' });
      }
    });

    watcher.on('add', name => {
      debug(`add file: %s`, name);

      const currentHash = md5(
        fs.readFileSync(path.join(appDirectory, name), 'utf8'),
      );

      hashMap.set(name, currentHash);

      hooksRunner.fileChange({ filename: name, eventType: 'add' });
    });

    watcher.on('unlink', name => {
      debug(`remove file: %s`, name);

      if (hashMap.has(name)) {
        hashMap.delete(name);
      }

      hooksRunner.fileChange({ filename: name, eventType: 'unlink' });
    });

    watcher.on('error', err => {
      throw err;
    });

    return watcher;
  }
};
