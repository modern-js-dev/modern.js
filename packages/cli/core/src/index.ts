import path from 'path';
import {
  compatRequire,
  pkgUp,
  ensureAbsolutePath,
  logger,
} from '@modern-js/utils';
import {
  createAsyncManager,
  createAsyncWorkflow,
  createParallelWorkflow,
  ParallelWorkflow,
  AsyncWorkflow,
  Progresses2Runners,
} from '@modern-js/plugin';
import { enable } from '@modern-js/plugin/node';

import type { Hooks } from '@modern-js/types';
import { program, Command } from './utils/commander';
import {
  resolveConfig,
  defineConfig,
  loadUserConfig,
  UserConfig,
  ToolsConfig,
} from './config';
import { loadPlugins } from './loadPlugins';
import {
  AppContext,
  ConfigContext,
  IAppContext,
  initAppContext,
  ResolvedConfigContext,
  useAppContext,
  useConfigContext,
  useResolvedConfigContext,
} from './context';
import { initWatcher } from './initWatcher';
import { NormalizedConfig } from './config/mergeConfig';
import { loadEnv } from './loadEnv';

export type { Hooks };
export { defaultsConfig } from './config';

export * from '@modern-js/plugin';
export * from '@modern-js/plugin/node';

program
  .name('modern')
  .usage('<command> [options]')
  .version(process.env.MODERN_JS_VERSION || '0.1.0');

export type HooksRunner = Progresses2Runners<{
  config: ParallelWorkflow<void>;
  validateSchema: ParallelWorkflow<void>;
  prepare: AsyncWorkflow<void, void>;
  commands: AsyncWorkflow<
    {
      program: Command;
    },
    void
  >;
  watchFiles: ParallelWorkflow<void>;
  fileChange: AsyncWorkflow<
    {
      filename: string;
    },
    void
  >;
  beforeExit: AsyncWorkflow<void, void>;
}>;

const hooksMap = {
  config: createParallelWorkflow(),
  validateSchema: createParallelWorkflow(),
  // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
  prepare: createAsyncWorkflow<void, void>(),
  commands: createAsyncWorkflow<
    {
      program: Command;
    },
    // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
    void
  >(),
  watchFiles: createParallelWorkflow(),
  fileChange: createAsyncWorkflow<
    {
      filename: string;
    },
    // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
    void
  >(),
  // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
  beforeExit: createAsyncWorkflow<void, void>(),
};

export const manager = createAsyncManager<Hooks, typeof hooksMap>(hooksMap);

export const {
  createPlugin,
  registe: registerHook,
  useRunner: mountHook,
} = manager;

export const usePlugins = (plugins: string[]) =>
  plugins.forEach(plugin =>
    manager.usePlugin(compatRequire(require.resolve(plugin))),
  );

export {
  defineConfig,
  AppContext,
  useAppContext,
  useConfigContext,
  useResolvedConfigContext,
};

export type { NormalizedConfig, IAppContext, UserConfig, ToolsConfig };

const initAppDir = async (): Promise<string> => {
  const pkg = await pkgUp({ cwd: process.cwd() });

  if (!pkg) {
    throw new Error(
      `no package.json found in current work dir: ${process.cwd()}`,
    );
  }

  return path.dirname(pkg);
};

export interface CoreOption {
  dryRun?: boolean;
}

const createCli = () => {
  let hooksRunner: HooksRunner;
  let isRestart = false;

  const init = async (argv: string[] = []) => {
    enable();

    manager.clear();

    const appDirectory = await initAppDir();

    loadEnv(appDirectory);

    const loaded = await loadUserConfig(appDirectory);

    const plugins = loadPlugins(appDirectory, loaded.config.plugins || []);

    plugins.forEach(plugin => plugin.cli && manager.usePlugin(plugin.cli));

    const appContext = initAppContext(appDirectory, plugins, loaded.filePath);

    manager.run(() => {
      ConfigContext.set(loaded.config);
      AppContext.set(appContext);
    });

    hooksRunner = await manager.init();

    ['SIGINT', 'SIGTERM', 'unhandledRejection', 'uncaughtException'].forEach(
      event => {
        process.on(event, async err => {
          await hooksRunner.beforeExit();
          if (err instanceof Error) {
            logger.error(err.stack);
          }
          process.nextTick(() => {
            // eslint-disable-next-line no-process-exit
            process.exit(1);
          });
        });
      },
    );

    const extraConfigs = await hooksRunner.config();

    const extraSchemas = await hooksRunner.validateSchema();

    const resolved = await resolveConfig(
      loaded,
      extraConfigs as any,
      extraSchemas as any,
      isRestart,
      argv,
    );

    // update context value
    manager.run(() => {
      ConfigContext.set(loaded.config);
      ResolvedConfigContext.set(resolved);
      AppContext.set({
        ...appContext,
        port: resolved.server.port!,
        distDirectory: ensureAbsolutePath(appDirectory, resolved.output.path!),
      });
    });

    await hooksRunner.prepare();

    return { loadedConfig: loaded, appContext, resolved };
  };

  async function run(argv: string[]) {
    const { loadedConfig, appContext, resolved } = await init(argv);

    await hooksRunner.commands({ program });

    initWatcher(
      loadedConfig,
      appContext.appDirectory,
      resolved.source.configDir,
      hooksRunner,
      argv,
    );
    manager.run(() => program.parse(process.argv));
  }

  async function restart() {
    isRestart = true;

    logger.info('Restart...\n');

    let hasGetError = false;
    try {
      await init(process.argv.slice(2));
    } catch (err) {
      console.error(err);
      hasGetError = true;
    } finally {
      if (!hasGetError) {
        manager.run(() => program.parse(process.argv));
      }
    }
  }

  return {
    init,
    run,
    restart,
  };
};

export const cli = createCli();
