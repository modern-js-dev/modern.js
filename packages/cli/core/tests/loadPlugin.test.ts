import path from 'path';
import { loadPlugins, getAppPlugins } from '../src/loadPlugins';

describe('load plugins', () => {
  test('getAppPlugins', () => {
    const appDirectory = path.resolve(
      __dirname,
      './fixtures/load-plugin/user-plugins',
    );
    const plugins = getAppPlugins(appDirectory, ['foo' as any], {
      x: {
        cli: 'x',
        forced: true,
      } as any,
    });
    expect(plugins).toEqual([{ cli: 'x', forced: true }, 'foo']);
  });

  test('should load user plugin successfully', () => {
    const fixture = path.resolve(
      __dirname,
      './fixtures/load-plugin/user-plugins',
    );

    const plugins = loadPlugins(fixture, {
      plugins: [
        { cli: path.join(fixture, './test-plugin-a.js') },
        { server: './test-plugin-b' },
      ],
    });

    expect(plugins).toEqual([
      {
        cli: {
          name: 'a',
          pluginPath: path.join(fixture, './test-plugin-a.js'),
        },
        cliPkg: path.join(fixture, './test-plugin-a.js'),
      },
      {
        server: {
          pluginPath: path.join(fixture, './test-plugin-b.js'),
        },
        serverPkg: './test-plugin-b',
      },
    ]);
  });

  test('should pass options to Plugin', () => {
    const fixture = path.resolve(
      __dirname,
      './fixtures/load-plugin/user-plugins',
    );

    const plugins = loadPlugins(fixture, {
      plugins: [{ cli: ['./test-plugin-c', 'c'] }, ['./test-plugin-c', 'c2']],
    });

    expect(plugins).toMatchSnapshot();
  });

  test('should load user string plugin successfully', () => {
    const fixture = path.resolve(
      __dirname,
      './fixtures/load-plugin/user-plugins',
    );

    const plugins = loadPlugins(fixture, {
      plugins: [path.join(fixture, './test-plugin-a.js') as any],
    });

    expect(plugins).toEqual([
      {
        cli: {
          name: 'a',
          pluginPath: path.join(fixture, './test-plugin-a.js'),
        },
        cliPkg: path.join(fixture, './test-plugin-a.js'),
      },
    ]);
  });

  test(`should throw error when plugin not found `, () => {
    const fixture = path.resolve(__dirname, './fixtures/load-plugin/not-found');

    expect(() => {
      loadPlugins(fixture, {
        plugins: [{ cli: './test-plugin-a' }, { cli: './plugin-b' }],
      });
    }).toThrowError(/^Can not find plugin /);
  });
});
