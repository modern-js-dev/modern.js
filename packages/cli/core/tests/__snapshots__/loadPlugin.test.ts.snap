// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load plugins should pass options to Plugin 1`] = `
Array [
  Object {
    "cli": Object {
      "ASYNC_PLUGIN_SYMBOL": "ASYNC_PLUGIN_SYMBOL",
      "name": "c",
      "pluginPath": "/Users/chenjiahan/Project/modern.js/packages/cli/core/tests/fixtures/load-plugin/user-plugins/test-plugin-c.js",
      "post": Array [],
      "pre": Array [],
      "registerHook": Object {},
      "required": Array [],
      "rivals": Array [],
      "setup": [Function],
      "usePlugins": Array [],
    },
    "cliPkg": "./test-plugin-c",
  },
  Object {
    "cli": Object {
      "ASYNC_PLUGIN_SYMBOL": "ASYNC_PLUGIN_SYMBOL",
      "name": "c2",
      "pluginPath": "/Users/chenjiahan/Project/modern.js/packages/cli/core/tests/fixtures/load-plugin/user-plugins/test-plugin-c.js",
      "post": Array [],
      "pre": Array [],
      "registerHook": Object {},
      "required": Array [],
      "rivals": Array [],
      "setup": [Function],
      "usePlugins": Array [],
    },
    "cliPkg": "./test-plugin-c",
  },
]
`;
